### PR TITLE
Use signal-cli-dbus-rest-api over abandoned signal-web-gateway.

### LINKED ISSUE #478: support for https://github.com/bbernhard/signal-cli-rest-api
there is a nice working stable project for signal gateway. can this also be supported ? 

https://github.com/bbernhard/signal-cli-rest-api

thanks

### PR DESCRIPTION
**Pull request by rotdrop**
See https://gitlab.com/morph027/signal-cli-dbus-rest-api.

This would fix  #478.

I have read #454

### CODE REVIEW

**Comment by rotdrop** [comment]
The test-suite needs to be updated:

```
This version of Nextcloud requires at least PHP 7.4<br/>You are currently running 7.3.33-1+ubuntu20.04.1+deb.sury.org+1. Please update your PHP version.
```

File: lib/Service/Gateway/Signal/Gateway.php [hunk 1]
```diff
@@ -62,21 +62,41 @@ public function __construct(IClientService $clientService,
...
-			[
-				'body' => [
-					'to' => $identifier,
-					'message' => $message,
-				],
-			]
-		);
-		$body = $response->getBody();
-		$json = json_decode($body, true);
+		// determine type of gateway
```
Comment (vitormattos) [inline lib/Service/Gateway/Signal/Gateway.php:65]: Is this comment necessary?

File: lib/Service/Gateway/Signal/Gateway.php [hunk 2]
```diff
@@ -62,21 +62,41 @@ public function __construct(IClientService $clientService,
...
+		$response = $client->get($this->config->getUrl() . '/v1/about');
+		if ($response->getStatusCode() === 200) {
+			// New style gateway https://gitlab.com/morph027/signal-cli-dbus-rest-api
```
Comment (vitormattos) [inline lib/Service/Gateway/Signal/Gateway.php:68]: Is this comment necessary?

File: lib/Service/Gateway/Signal/Gateway.php [hunk 3]
```diff
@@ -62,21 +62,41 @@ public function __construct(IClientService $clientService,
...
+				]
+			);
+			$body = $response->getBody();
+			$json = json_decode($body, true);
+			if ($response->getStatusCode() !== 201 || is_null($json) || !is_array($json) || !isset($json['timestamp'])) {
+				$status = $response->getStatusCode();
+				throw new SmsTransmissionException("error reported by Signal gateway, status=$status, body=$body}");
+			}
+		} else {
+			// Try old deprecated gateway https://gitlab.com/morph027/signal-web-gateway
```
Comment (vitormattos) [inline lib/Service/Gateway/Signal/Gateway.php:82]: Is this comment necessary?

File: lib/Service/SetupService.php [hunk 4]
```diff
@@ -91,7 +91,7 @@ public function startSetup(IUser $user, string $gatewayName, string $identifier)
 		try {
 			$gateway->send($user, $identifier, "$verificationNumber is your Nextcloud verification code.");
 		} catch (SmsTransmissionException $ex) {
-			throw new VerificationTransmissionException('could not send verification code');
+			throw new VerificationTransmissionException('could not send verification code', $ex->getCode(), $ex);
```
Comment (vitormattos) [inline lib/Service/SetupService.php:94]: Have a place of code using the 2th and 3th argument?

**Pull request merged**